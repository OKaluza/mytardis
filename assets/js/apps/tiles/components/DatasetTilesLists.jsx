import React, { useEffect, useState } from 'react';
import ReactDOM from 'react-dom';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import PropTypes from "prop-types";
import { fetchDatasetsForExperiment, fetchExperimentList, shareDataset } from './utils/FetchData';
import DatasetTiles from './DatasetTiles';
import ExperimentListDropDown from './SelectExperiment';

const DatasetTilesLists = ({ shareContainer, experimentId }) => {
  const [mainListData, setMainListData] = useState([]);
  const [shareListData, setShareListData] = useState([]);
  const [expListValue, setExpListValue] = useState();
  const onDragEnd = (result) => {
    // dropped nowhere
    if (!result.destination) {
      return;
    }
    // dropped in the same list
    if (result.destination.droppableId === result.source.droppableId) {
      return;
    }
    // dropped from share to main list
    if (result.destination.droppableId === 'main-list'
    && result.source.droppableId === 'share-list') {
      // get body data
      const data = shareListData[result.source.index];
      // update experiment dataset
      const datasetId = result.draggableId.split('_')[1];
      shareDataset(JSON.stringify(data), experimentId, datasetId)
        .then(() => {
          // fetch data and update main list
          fetchDatasetsForExperiment(experimentId)
            .then(listData => setMainListData(listData));
        });
    }
    // dropped from main to share list
    if (result.destination.droppableId === 'share-list'
    && result.source.droppableId === 'main-list') {
      // get body data
      const data = mainListData[result.source.index];
      // update experiment dataset
      const datasetId = result.draggableId.split('_')[1];
      shareDataset(JSON.stringify(data), expListValue, datasetId)
        .then(() => {
          // fetch data and update share list
          fetchDatasetsForExperiment(expListValue)
            .then(listData => setShareListData(listData));
        });
    }
  };
  const onChange = (event) => {
    event.preventDefault();
    setExpListValue(event.target.value);
    fetchDatasetsForExperiment(event.target.value).then(result => setShareListData(result));
  };
  useEffect(() => {
    fetchDatasetsForExperiment(experimentId).then(result => setMainListData(result));
    // load initial list
    fetchExperimentList().then((expList) => {
      // TODO
      setExpListValue(expList[0].id);
      fetchDatasetsForExperiment(expList[0].id).then(result => setShareListData(result));
    });
  }, [experimentId]);
  return (
    <DragDropContext onDragEnd={onDragEnd}>
      <Droppable droppableId="main-list">
        {provided => (
          <div ref={provided.innerRef} {...provided.droppableProps}>
            <DatasetTiles data={mainListData} listName="main-list" />
            {provided.placeholder}
          </div>
        )}
      </Droppable>
      {ReactDOM.createPortal(
        <Droppable droppableId="share-list">
          {provided => (
            <div ref={provided.innerRef} {...provided.droppableProps}>
              <ExperimentListDropDown onChange={onChange} value={expListValue} />
              <DatasetTiles data={shareListData} listName="share-list" />
              {provided.placeholder}
            </div>
          )}
        </Droppable>, shareContainer,
      )}
    </DragDropContext>
  );
};
DatasetTilesLists.propTypes = {
  shareContainer: PropTypes.object.isRequired,
  experimentId: PropTypes.string.isRequired,
};
export default DatasetTilesLists;
