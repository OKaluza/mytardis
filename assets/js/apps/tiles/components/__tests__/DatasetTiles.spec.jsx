import React from 'react';
import ReactDOM from 'react-dom';
import { act } from '@testing-library/react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import DatasetTilesLists from '../DatasetTilesLists';

const fakeData = [
  {
    id: 26,
    description: 'My demo dataset 25062019',
    directory: null,
    created_time: '2019-06-25 01:54:04.864000+00:00',
    modified_time: '2020-04-22 02:03:17.802951+00:00',
    immutable: false,
    instrument: null,
    experiments: [
      34,
    ],
    file_count: 1,
    url: '/dataset/26',
    size: 164495,
    size_human_readable: '160.6 KB',
    show_dataset_thumbnails: true,
  },
  {
    id: 30,
    description: 'New Dataset',
    directory: null,
    created_time: '2020-01-13 01:36:34.761466+00:00',
    modified_time: '2020-04-22 02:47:58.368048+00:00',
    immutable: false,
    instrument: null,
    experiments: [
      34,
    ],
    file_count: 1,
    url: '/dataset/30',
    size: 350565,
    size_human_readable: '342.3 KB',
    show_dataset_thumbnails: true,
  }];

global.fetch = require('jest-fetch-mock');

configure({ adapter: new Adapter() });


let container = null;
let shareContainer = null
let component = null;
beforeEach(async () => {
  jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
    json: () => Promise.resolve(fakeData),
  }));
  container = document.createElement('div');
  document.body.appendChild(container);
  shareContainer = document.createElement('div');
  document.body.appendChild(shareContainer);
  await act(async () => {
    component = mount(<DatasetTilesLists experimentId="1234" shareContainer={shareContainer} />, { attachTo: container });
  });
  component.update();
});
afterEach(() => {
  // cleanup on exiting
  container.remove();
  container = null;
  component.unmount();
  component.detach();
});
describe('render dataset tiles on page load', () => {
  it('should match snapshot', async () => {
    await act(async () => {
      ReactDOM.render(<DatasetTilesLists experimentId={'1234'} shareContainer={shareContainer} />, container);
    });
    expect(container).toMatchSnapshot();
  });
});
