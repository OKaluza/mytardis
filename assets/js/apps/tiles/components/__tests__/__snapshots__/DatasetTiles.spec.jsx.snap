// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render dataset tiles on page load should match snapshot 1`] = `
<div
  id="main-list"
>
  <div
    data-rbd-droppable-context-id="0"
    data-rbd-droppable-id="main-list"
  >
    <ul
      class="datasets thumbnails"
    >
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="main-list_26"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="main-list_26"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <li
          class="dataset card mb-2"
        >
          <div
            class="card-body"
          >
            <div
              class="float-left"
              style="margin-right: 10px;"
            >
              <div
                style="display: flex; text-align: center; font-size: 32px;"
              >
                <i
                  class="fa fa-folder-open o-6"
                />
              </div>
              <br />
              <label
                class="label label-info"
                for="dataset-checkbox-26"
                title="Mark dataset for download"
              >
                <i
                  class="fa fa-download"
                />
                <input
                  class="dataset_checkbox ml-2"
                  id="dataset-checkbox-26"
                  name="dataset"
                  style="display: inline;"
                  type="checkbox"
                  value="26"
                />
              </label>
            </div>
            <div
              class="float-left"
              style="margin-right: 10px;"
            />
            <div
              class="float-right"
              style="text-align: right;"
            >
              <p>
                <span
                  class="badge badge-info"
                  title="1 file"
                >
                  <i
                    class="fa fa-file"
                  />
                   
                  1
                </span>
              </p>
              <p>
                <span
                  class="badge badge-info"
                  title="Dataset size is"
                >
                  160.64 KB
                </span>
              </p>
            </div>
            <div
              class="float-left ds-thumb"
              style="margin-right: 10px;"
            >
              <canvas
                style="width: 100px;"
              />
            </div>
            <div
              class="float-left"
            >
              <p
                class="mb-2"
              >
                <a
                  class="dataset-link"
                  href="/dataset/26"
                >
                  My demo dataset 25062019
                </a>
              </p>
              
              <div
                style="clear: both;"
              />
            </div>
          </div>
        </li>
      </div>
      <div
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="main-list_30"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="main-list_30"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <li
          class="dataset card mb-2"
        >
          <div
            class="card-body"
          >
            <div
              class="float-left"
              style="margin-right: 10px;"
            >
              <div
                style="display: flex; text-align: center; font-size: 32px;"
              >
                <i
                  class="fa fa-folder-open o-6"
                />
              </div>
              <br />
              <label
                class="label label-info"
                for="dataset-checkbox-30"
                title="Mark dataset for download"
              >
                <i
                  class="fa fa-download"
                />
                <input
                  class="dataset_checkbox ml-2"
                  id="dataset-checkbox-30"
                  name="dataset"
                  style="display: inline;"
                  type="checkbox"
                  value="30"
                />
              </label>
            </div>
            <div
              class="float-left"
              style="margin-right: 10px;"
            />
            <div
              class="float-right"
              style="text-align: right;"
            >
              <p>
                <span
                  class="badge badge-info"
                  title="1 file"
                >
                  <i
                    class="fa fa-file"
                  />
                   
                  1
                </span>
              </p>
              <p>
                <span
                  class="badge badge-info"
                  title="Dataset size is"
                >
                  342.35 KB
                </span>
              </p>
            </div>
            <div
              class="float-left ds-thumb"
              style="margin-right: 10px;"
            >
              <canvas
                style="width: 100px;"
              />
            </div>
            <div
              class="float-left"
            >
              <p
                class="mb-2"
              >
                <a
                  class="dataset-link"
                  href="/dataset/30"
                >
                  New Dataset
                </a>
              </p>
              
              <div
                style="clear: both;"
              />
            </div>
          </div>
        </li>
      </div>
    </ul>
  </div>
</div>
`;
